### PR TITLE
[API] Crypto mode 'auto' implemented for listener.

### PR DESCRIPTION
**Pull request by maxsharabayko**
For an SRT listener to accept any crypto mode (AES-GCM or AES-CTR) the `SRTO_CRYPTOMODE=0` (Auto) can be used.

When `SRTO_CRYPTOMODE=0` (Auto) is set on an SRT Caller or SRT Rendezvous the actual mode to be negotiated is AES-CTR.

This change does not require further HS changes on the wire, still allows to let the listener to accept any crypto mode.

Once a connection is established, the actual crypto mode is returned via the `SRTO_CRYPTOMODE`: 1 (AES-CTR) or 2 (AES-GCM).

### CODE REVIEW

File: docs/API/API-socket-options.md [hunk 1]
```diff
@@ -333,12 +333,42 @@ The encryption mode to be used if the [`SRTO_PASSPHRASE`](#SRTO_PASSPHRASE) is s
...
-- `0`: auto-select during handshake negotiation (to be implemented; currently similar to AES-CTR).
+- `0`: SRT listener accepts any mode from the caller. SRT Caller or SRT Rendezvous effectively negotiate AES-CTR (1).
 - `1`: regular AES-CTR (without message integrity authentication).
 - `2`: AES-GCM mode with message integrity authentication (AEAD).
 
 The AES-GCM mode (2) is only allowed if TSBPD is enabled ([`SRTO_TSBPDMODE`](#SRTO_TSBPDMODE)).
-
+In rendezvous and for a caller Auto (0) is equivalent to AES-CTR (1).
+
+Once connection is established, reading the `SRTO_CRYPTOMODE` shows the negotiated crypto mode:
```
Comment (stevomatthews) [inline docs/API/API-socket-options.md:343]:
```suggestion
Once a connection is established, reading `SRTO_CRYPTOMODE` shows the negotiated crypto mode:
```

File: docs/API/API-socket-options.md [hunk 2]
```diff
@@ -333,12 +333,42 @@ The encryption mode to be used if the [`SRTO_PASSPHRASE`](#SRTO_PASSPHRASE) is s
...
+AES-CTR (1) or AES-GCM (2).
+
+When the [listener callback](./API-functions.md##srt_listen_callback) is used, the `SRTO_CRYPTOMODE`
+read on the new SRT socket to be accepted is not yet the negotiated but, but the one to negotiate,
+and inherited from the listener SRT socket.
```
Comment (stevomatthews) [inline docs/API/API-socket-options.md:348]:
```suggestion
When [Listener callback](./API-functions.md##srt_listen_callback) is used, the value of `SRTO_CRYPTOMODE`
read on the new SRT socket to be accepted is not yet the negotiated one. It is the value to be negotiated, and is inherited from the listener SRT socket.
```

File: docs/API/API-socket-options.md [hunk 3]
```diff
@@ -333,12 +333,42 @@ The encryption mode to be used if the [`SRTO_PASSPHRASE`](#SRTO_PASSPHRASE) is s
...
+If a specific behavior for each individual connection request is desired based on
+[the user ID or anything else](../features/access-control.md),
+the indended behavior can be achieved by setting the `SRTO_CRYPTOMODE` on the new SRT socket to a specific value.
```
Comment (stevomatthews) [inline docs/API/API-socket-options.md:351]:
```suggestion
the intended behavior can be achieved by setting the `SRTO_CRYPTOMODE` on the new SRT socket to a specific value.
```

File: docs/API/API-socket-options.md [hunk 4]
```diff
@@ -333,12 +333,42 @@ The encryption mode to be used if the [`SRTO_PASSPHRASE`](#SRTO_PASSPHRASE) is s
...
+
+For example, let's say the initial value set on the listener socket is Auto (0). The listener callback is called
+signalling the new connection request. The user ID is extracted, and the server wants to force AES-GCM only
+for connection from this very user. Then AES-GCM (2) can be set on the new socket.
```
Comment (stevomatthews) [inline docs/API/API-socket-options.md:355]:
```suggestion
For example, let's say the initial value set on the listener socket is Auto (0). The listener callback is called,
signalling the new connection request. The user ID is extracted, and the server wants to force AES-GCM only
for the connection from this specific user. In this case AES-GCM (2) can be set on the new socket.
```

File: docs/API/API-socket-options.md [hunk 5]
```diff
@@ -333,12 +333,42 @@ The encryption mode to be used if the [`SRTO_PASSPHRASE`](#SRTO_PASSPHRASE) is s
...
+
+There is no way to check the crypto mode being requested by the SRT caller at this point.
+
+| Caller      | Listener    | Negotiated  | | Rdv In-tor  | Rdv Res-der | Negotiated  |
+| ----------- | ----------- | ----------- |-| ----------- | ----------- | ----------- |
```
Comment (stevomatthews) [inline docs/API/API-socket-options.md:360]:
```suggestion
|             |             |             | | Rendezvous  | Rendezvous  | Negotiated  |
| Caller      | Listener    | Negotiated  | | Initiator   | Responder   | Negotiated  |
| ----------- | ----------- | ----------- |-| ----------- | ----------- | ----------- |
```

File: docs/API/API-socket-options.md [hunk 6]
```diff
@@ -333,12 +333,42 @@ The encryption mode to be used if the [`SRTO_PASSPHRASE`](#SRTO_PASSPHRASE) is s
...
+| 0 (auto)    | AES-CTR (1) | AES-CTR (1) | | 0 (auto)    | AES-CTR (1) | AES-CTR (1) |
+| 0 (auto)    | AES-GCM (2) | reject      | | 0 (auto)    | AES-GCM (2) | reject      |
+| AES-CTR (1) | 0 (auto)    | AES-CTR (1) | | AES-CTR (1) | 0 (auto)    | AES-CTR (1) |
+| AES-CTR (1) | AES-CTR (1) | AES-CTR (1) | | AES-CTR (1) | AES-CTR (1) | AES-CTR (1) |
+| AES-CTR (1) | AES-GCM (2) | reject      | | AES-CTR (1) | AES-GCM (2) | reject      |
+| AES-GCM (2) | 0 (auto)    | AES-GCM (2) | | AES-GCM (2) | 0 (auto)    | reject      |
+| AES-GCM (2) | AES-CTR (1) | reject      | | AES-GCM (2) | AES-CTR (1) | reject      |
+| AES-GCM (2) | AES-GCM (2) | AES-GCM (2) | | AES-GCM (2) | AES-GCM (2) | AES-GCM (2) |
+
+* Rdv - Rendezvous; In-tor - initiator; Res-der - responder.
```
Comment (stevomatthews) [inline docs/API/API-socket-options.md:371]: Not needed if you accept my suggested headings.

File: srtcore/core.cpp [hunk 7]
```diff
@@ -3433,6 +3436,10 @@ void srt::CUDT::startConnect(const sockaddr_any& serv_addr, int32_t forced_isn)
     // handle handshake extension flags.
     m_ConnReq.m_iType = UDT_DGRAM;
 
+    // Auto mode for caller and in rendezvous is equivalent to CIPHER_MODE_AES_CTR.
```
Comment (stevomatthews) [inline srtcore/core.cpp:3439]:
```suggestion
    // Auto mode for Caller and in Rendezvous is equivalent to CIPHER_MODE_AES_CTR.
```

File: docs/API/API-socket-options.md (hunk 6 above)
Comment (maxsharabayko) [reply docs/API/API-socket-options.md:371]: I wish, but the third row from the bottom is different. See the images in the description of the issue for better visibility.

**Pull request merged**